### PR TITLE
Fix: Unexpected behaviour: parse_numbers() doesn't handle space as a grouping symbol, however, parse_decimal() does #1061

### PR DESCRIPTION
**Pull request by sebas-inf**
Cleaned the string to ensure that it still works if there is a non-breakable space, \xa0, in the number and separed the logic so that it is easeir to understand.

### CODE REVIEW

**Review by akx — CHANGES_REQUESTED** [review]
You've committed and pushed your entire `.venv` – please clean that up first.

**Comment by sebas-inf** [comment]
> You've committed and pushed your entire `.venv` – please clean that up first.

Should I just push the file I changed or everything except the .venv?

**Comment by akx** [comment]
> Should I just push the file I changed or everything except the .venv?

Your commit(s) should only have relevant changes.

**Comment by sebas-inf** [comment]
> > Should I just push the file I changed or everything except the .venv?
> 
> Your commit(s) should only have relevant changes.

Ok, I will push the file with the fix only then.

**Comment by sebas-inf** [comment]
> You still have 755 files in this pull request. You'll probably need to `git reset 40e60a1f6cf178d9f57fcc14f157ea1b2ab77361` (the current `master`), then carefully `git add` only the changes you want, `git commit` them, and `git push --force-with-lease` them to your repository.

should i do all of this within the terminal in vscode where i was working?

**Comment by akx** [comment]
> should i do all of this within the terminal in vscode where i was working?

Maybe. I don't use Visual Studio Code much. Visual Studio Code might have other tools for backing out commits.

**Comment by sebas-inf** [comment]
> > should i do all of this within the terminal in vscode where i was working?
> 
> Maybe. I don't use Visual Studio Code much. Visual Studio Code might have other tools for backing out commits.

It worked, I made it so i only pushed the numbers.py file, how do i change that in this pr now?

File: babel/numbers.py [hunk 1]
```diff
@@ -1027,11 +1029,21 @@ def parse_number(
     :raise `UnsupportedNumberingSystemError`: if the numbering system is not supported by the locale.
     """
     try:
-        return int(string.replace(get_group_symbol(locale, numbering_system=numbering_system), ''))
+        # Get the group symbol from the locale
+        group_symbol = get_group_symbol(locale, numbering_system=numbering_system)
+
+        # Replace non-breakable spaces with the group symbol
+        cleaned_string = string.replace('\xa0', group_symbol)
```
Comment (akx) [inline babel/numbers.py:1036]: Could you add a test case (either a doctest or a real test in `tests/`) that uses an `\xa0` character?

**Review by akx — CHANGES_REQUESTED** [review]
I would like to err on the side of caution here, if there is the remotest chance this could cause mis-parsing of numbers (e.g. something that couldn't be parsed before being now accidentally parsed as something it's not intended to be parsed as).

To that end, maybe this function should also have a `strict=True` parameter like `parse_decimal` does, and then maybe the grouping character logic could be shared between the two in non-strict mode?

File: babel/numbers.py (hunk 1 above)
Comment (sebas-inf) [reply babel/numbers.py:1036]: would this be a valid test case for that: >>> parse_number('1 099', locale='ru')

Comment (sebas-inf) [reply babel/numbers.py]: made some changes applying some of the logic from parse_decimal and it says its passing the test cases on my end

File: babel/numbers.py [hunk 2]
```diff
@@ -1001,6 +1001,7 @@ def __init__(self, message: str, suggestions: list[str] | None = None) -> None:
 def parse_number(
     string: str,
     locale: Locale | str | None = LC_NUMERIC,
+    strict: bool = False,
```
Comment (akx) [inline babel/numbers.py:1004]: Please add the new argument into the docstring. `strict` mode should probably be the default here, and the new behavior (which allows more loose strings to be parsed) should be opt-in.

File: babel/numbers.py [hunk 3]
```diff
@@ -1010,6 +1011,8 @@ def parse_number(
     1099
     >>> parse_number('1.099', locale='de_DE')
     1099
+    >>> parse_number('1 099', locale='ru')
```
Comment (akx) [inline babel/numbers.py:1014]: If the intent is to test a string that has a non-breaking space character, make it obvious:
```suggestion
    >>> parse_number('1\xa0099', locale='ru')
```

File: babel/numbers.py [hunk 4]
```diff
@@ -1027,11 +1030,67 @@ def parse_number(
...
+                        )
+                    else:
+                        raise NumberFormatError(
+                            f"{string!r} is not a properly formatted decimal number. "
+                            f"Did you mean {proper!r}? Or maybe {proper_alt!r}?",
+                            suggestions=[proper, proper_alt],
+                        )
+
+        # Remove other spaces and replace the group symbol with an empty string
+        cleaned_string = string.replace(' ', '').replace(group_symbol, '')
```
Comment (akx) [inline babel/numbers.py:1085]: `format_number` is not meant to parse decimal numbers, only ever integers, anything decimal-related shouldn't be here at all.

If anything, we should check for the existence of a decimal separator in the input in strict mode, and fail if it is there.

File: babel/numbers.py [hunk 5]
```diff
@@ -1027,11 +1030,67 @@ def parse_number(
...
+        # Get the group symbol from the locale
+        group_symbol = get_group_symbol(locale, numbering_system=numbering_system)
+
+        # Replace non-breakable spaces with the group symbol in non-strict mode
+        if not strict and group_symbol == '\xa0' and '\xa0' not in string and ' ' in string:
+            string = string.replace(' ', group_symbol)
+
+        # If strict mode is enabled, handle irregular formatting
+        if strict and group_symbol in string:
+            # Your additional logic for strict mode here
```
Comment (akx) [inline babel/numbers.py:1042]: What a strange comment... 😁

**Review by akx — CHANGES_REQUESTED** [review]

File: babel/numbers.py (hunk 4 above)
Comment (sebas-inf) [reply babel/numbers.py]: made the changes in the latest commit

**Comment by akx** [comment]
@sebas-inf If you're still around, please see the comments and rebase this to resolve conflicts. Thanks :)

**Pull request closed without merging**